KAFKA-12343: Handle exceptions better in TopicAdmin, including UnsupportedVersionException

rhauch opened the pull request:
Refactored the KafkaBasedLog logic to read end offsets into a separate method to make it easier to test. Also changed the TopicAdmin.endOffsets method to throw the original UnsupportedVersionException, LeaderNotAvailableException, and TimeoutException rather than wrapping, to better conform with the consumer method and how the KafkaBasedLog retries those exceptions.

Added new tests to verify various scenarios and errors.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

rhauch committed:
KAFKA-12343: Handle exceptions better in TopicAdmin

Refactored the KafkaBasedLog logic to read end offsets into a separate method to make it easier to test. Also changed the TopicAdmin.endOffsets method to throw the original UnsupportedVersionException, LeaderNotAvailableException, and TimeoutException rather than wrapping, to better conform with the consumer method and how the KafkaBasedLog retries those exceptions.

Added new tests to verify various scenarios and errors.

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -390,7 +410,11 @@ public void run() {
-                                "This may occur when brokers are unavailable or unreachable. Reason: {}", topic, e.getMessage());
+                                     "This may occur when brokers are unavailable or unreachable. Reason: {}", topic, e.getMessage());
+                            continue;
+                        } catch (RetriableException | org.apache.kafka.connect.errors.RetriableException e) {
```
`readToLogEnd` is called by `start`. Should we add similar exception handle for that?

rhauch replied:
The previous `readToLogEnd()` that used the consumer did not have any special exception handling in the `start()` method. Any problem to start seems like it should propagate up to result in the worker failing. Especially when coupled with your fix in #10152.

I added this here because retriable exceptions should not stop the KafkaBasedLog's thread.

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -366,6 +348,44 @@ private void readToLogEnd() {
+            // Use the admin client to immediately find the end offsets for the assigned topic partitions.
+            // Unlike using the consumer
+            try {
+                endOffsets = admin.endOffsets(assignment);
```
How about `return admin.endOffsets(assignment);`

rhauch replied:
I wanted to avoid 3 separate return statements.

rhauch replied:
I did see a way of reducing the # of returns to 2, and removing one of the duplicated lines. Thanks!

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -321,29 +325,7 @@ private void readToLogEnd() {
         log.trace("Reading to end of offset log");
```
redundant log message. `readEndOffsets(Set<TopicPartition>)` has similar log.

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -651,6 +651,10 @@ public Config describeTopicConfig(String topic) {
+     * @throws TimeoutException if the offset metadata could not be fetched before the amount of time allocated
+     *         by {@code request.timeout.ms} expires, and this call can be retried
+     * @throws LeaderNotAvailableException if the leader was not available and this call can be retried
      * @throws RetriableException if a retriable error occurs, the operation takes too long, or the
```
`the operation takes too long` 
^^^^^^^^^^^^^^^^^^^^^^^ this message gets invalid since `TimeoutException` is not wrapped to `RetriableException` anymore.

chia7712 reviewed (commented):
@rhauch Thanks for this nice patch. It can resolve a issue I encountered #10152 :)

rhauch committed:
Removed unused import

rhauch committed:
Incorporate review feedback

rhauch commented:
Thanks for the quick review, @chia7712. I've incorporated several of your suggestions, and added comments/responses on the others.

BTW, a run of the relevant Connect distributed system tests passed with this branch: https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4390/

rhauch committed:
Incorporated another suggestion from the review

rhauch commented:
@chia7712, for what it's worth, I still think your #10152 PR is still an important fix.

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -366,6 +346,42 @@ private void readToLogEnd() {
+        // Since all we're trying to do here is get the end offset, we should use the supplied admin client
+        // (if available)
+        // (which prevents 'consumer.endOffsets(...)'
+        // from
```
Since we are moving this comment, probably it's a good idea to fix this incomplete sentence here too

chia7712 commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -366,6 +346,42 @@ private void readToLogEnd() {
+    Map<TopicPartition, Long> readEndOffsets(Set<TopicPartition> assignment) {
+        log.trace("Reading to end of offset log");
+
+        Map<TopicPartition, Long> endOffsets;
```
unused variable

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -161,6 +164,7 @@ public void start() {
 
         // Create the topic admin client and initialize the topic ...
         admin = topicAdminSupplier.get();   // may be null
+        useAdminForListOffsets = admin != null;
```
It's not obvious to me why we have to wait to call `topicAdminSupplier.get();` in `start`. 

I think we can call it in the constructor and not save the field `topicAdminSupplier` but keep only `admin` as a member field. Unfortunately it can't be declared `final` because it may be reset (set to `null`) in the exception handling below but still, using fewer member fields seems reasonable and does not reduce the usability in this class IMO.

rhauch replied:
I don't think the wait time matters whether we call `topicAdminSupplier.get()` in the constructor or here, because the constructor and the `start()` method are called synchronously from the Connect & MM2 startup process.

However, I do think there is a lot of value in following the existing pattern of not actually gathering, creating or using resources until the `start()` method. And we still need all of the fields, including the `useAdminForListOffset` boolean field that is set in the start method but also may be changed later on in the `readEndOffsets(...)` method if the admin is talking to a broker that doesn't support the end offsets method (e.g., 0.10.x). In fact, the latter is the primary error condition this PR is trying to fix.

So, I think it's better to keep the code as is. But please let me know if you still disagree.

rhauch replied:
Per the next comment, I've removed the `useAdminForListOffset` field to simplify things. However, I still think that it's more appropriate and follows existing convention for this class to call `topicAdminSupplier.get()` in the `start()` method.

kkonstantine replied:
My main observation here is that keeping initialization in the constructor allows us to declare read only fields as `final`. 
It also offers better exception safety, in the sense that when an object is constructed (the call to its constructor succeeds) then we know that more or less we have a functional instance of that class. 

I'm not ignoring the cases where the initialization of variables has to happen at a later time during the call of `start`. But hopefully these cases are minimal, or else `start` ends up being the actual constructor of objects like this one here. Which can be redundant and in some cases more risky. 

As I mentioned, since `admin` is not immutable here (can be set to `null` later to support connection with older brokers) I'm fine keeping this call in `start`. But I'd still recommend not pushing initializations outside the constructor if they can happen at the time of the object construction (as is the case with `admin`'s initial initialization here).

kkonstantine commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -366,6 +346,42 @@ private void readToLogEnd() {
+                // This may happen with really old brokers that don't support the auto topic creation
+                // field in metadata requests
+                log.debug("Reading to end of log offsets with consumer since admin client is unsupported: {}", e.getMessage());
+                useAdminForListOffsets = false;
```
Using `admin = null` here allows to GC the unused admin instance earlier, right?
Not a big gain, but also I don't see much benefit by using a variable such as `useAdminForListOffsets`

rhauch replied:
I'm not sure the GC will really be any better if we were to use `admin = null` here, since the supplier is in charge of the lifecycle of the `TopicAdmin` instance, and our current implementation is that the `SharedTopicAdmin` is closed only when the herder is closed.

I did originally do what you're suggesting though, but changed it since technically we could use the topic admin for something else. Granted, the current code doesn't need it for anything else. I thought that using the variable made it a bit clearer that the variable controlled the behavior of the `readEndOffsets(...)` method using or not using the admin client.

But I think you're main point is that this can be simpler by using `admin = null`, and I can't really argue with that.

kkonstantine replied:
Good point. As long as `SharedTopicAdmin` admin won't be recycled. That's negligible anyways, I was referring to the pattern mainly.

kkonstantine reviewed (commented):
Thanks @rhauch 
Looks good overall. 
Left a few comments suggesting a reduction of member fields. Not sure we should use extra fields for readability but lmk what you think.

rhauch committed:
More improvements as suggested by reviews

rhauch commented:
@chia7712 and @kkonstantine: thanks for the reviews. I've incorporated your recent feedback, with the exception of @kkonstantine's suggestion to move the `topicAdminSupplier.get()` from `start()` into the constructor, as mentioned above in one of the comment threads.

Please take another look. Thanks!

kkonstantine reviewed (approved):
Thanks @rhauch, LGTM

Regarding initialization I replied inline, but not a blocker here of course.
